lib/gemstash/gem_yanker.rb

```diff
@@ -0,0 +1,58 @@
+require "gemstash"
+
+module Gemstash
+  # Class that supports yanking a gem from the private repository of gems.
+  class GemYanker
+    include Gemstash::Env::Helper
+
+    # This error is thrown when yanking a non-existing gem name.
+    class UnknownGemError < StandardError
+    end
+
+    # This error is thrown when yanking a non-existing gem version.
+    class UnknownVersionError < StandardError
+    end
+
+    # This error is thrown when yanking an already yanked gem version.
+    class YankedVersionError < StandardError
+    end
+
+    def initialize(auth_key, gem_name, slug)
+      @auth_key = auth_key
+      @gem_name = gem_name
+      @slug = slug
+      @db_helper = Gemstash::DBHelper.new
+    end
+
+    def yank
+      check_auth
+      update_database
+      invalidate_cache
+    end
+
+  private
+
+    def check_auth
+      raise Gemstash::NotAuthorizedError, "Authorization key required" if @auth_key.to_s.strip.empty?
+      auth = Authorization[@auth_key]
+      raise Gemstash::NotAuthorizedError, "Authorization key is invalid" unless auth
+      raise Gemstash::NotAuthorizedError, "Authorization key doesn't have yank access" unless auth.yank?
+    end
+
+    def update_database
+      gemstash_env.db.transaction do
+        gem_id = @db_helper.find_rubygem_id(@gem_name)
+        raise UnknownGemError, "Cannot yank an unknown gem!" unless gem_id
+        full_name = "#{@gem_name}-#{@slug}"
+        version = @db_helper.find_version_by_full_name(full_name)
+        raise UnknownVersionError, "Cannot yank an unknown version!" unless version
```

Comment:
It seems like the search for the gem_id is only done to assert that that gem actually exists.

Given that we are talking about yanking a gem by name and version `full_name`, does it make sense to just check that we can find it by name and version?

The check for the gem_name seems to be there just for double checking, but I am not too sure how much it helps to the clarity of the method as then the gem_id is not used for anything else.

If you want to keep the check, then it can be something like

``` ruby
raise UnknownGemError ... unless @db_helper.find_rubygem_id(@gem_name)
```

BTW, does it make sense to extract the `full_name` to a method for building?